docs/assembly.rst

```diff
@@ -1,17 +1,117 @@
+#################################################
 
 .. index:: ! assembly, ! asm, ! evmasm
 
-Solidity defines an assembly language that can also be used without Solidity.
-This assembly language can also be used as "inline assembly" inside Solidity
-source code. We start with describing how to use inline assembly and how it
-differs from standalone assembly and then specify assembly itself.
+Julia is an intermediate language that can compile to various different backends
+(EVM 1.0, EVM 1.5 and eWASM are planned).
+Because of that, it is designed to be as featureless as possible.
+It can already be used for "inline assembly" inside Solidity and
+future versions of the Solidity compiler will even use Julia as intermediate
+language. It should also be easy to build high-level optimizer stages for Julia.
+
+The core components of Julia are functions, blocks, variables, literals,
+for-loops, switch-statements, expressions and assignments to variables.
+
+Julia in itself does not even provide operators. If the EVM is targeted,
+opcodes will be available as built-in functions, but they can be reimplemented
+if the backend changes.
+
+The following example program assumes that the EVM opcodes ``mul``, ``div``
+and ``mod`` are available either natively or as functions and computes exponentiation.
+
+.. code::
+    {
+        function power(base, exponent) -> (result)
+        {
+            switch exponent
+            0: { result := 1 }
+            1: { result := base }
+            default:
+            {
+                result := power(mul(base, base), div(exponent, 2))
+                switch mod(exponent, 2)
+                    1: { result := mul(base, result) }
+            }
+        }
+    }
+
+It is also possible to implement the same function using a for-loop
+instead of recursion. Here, we need the EVM opcodes ``lt`` (less-than)
+and ``add`` to be available.
+
+.. code::
+    {
+        function power(base, exponent) -> (result)
+        {
+            result := 1
+            for { let i := 0 } lt(i, exponent) { i := add(i, 1) }
+            {
+                result := mul(result, base)
+            }
+        }
+    }
+
+Specification of Julia
+======================
+
+Grammar::
+
+    Block = '{' Statement* '}'
+    Statement =
+        Block |
+        FunctionDefinition |
+        VariableDeclaration |
+        Assignment |
+        Expression |
+        Switch |
+        ForLoop |
+        'break' | 'continue'
+        SubAssembly
+    FunctionDefinition =
+        'function' Identifier '(' IdentifierList? ')'
+        ( '->' '(' IdentifierList ')' )? Block
+    VariableDeclaration =
+        'let' IdentifierOrList ':=' Expression
+    Assignment =
+        IdentifierOrList ':=' Expression
+    Expression =
+        Identifier | Literal | FunctionCall
+    Switch =
+        'switch' Expression Case* ( 'default' ':' Block )?
+    Case =
+        'case' Expression ':' Block
+    ForLoop =
+        'for' Block Expression Block Block
```

Comment:
The reason for all this are the complicated scoping rules for the for-loop. We want to allow a new variable to be declared whose scope is only the for-loop itself, so basically the idiomatic

`for { let x := 1 } lt(x, 10) { x := add(x, 1) } { ...}`

Of course it is weird that the scope of `x` extends beyond `{ let x := 1 }` but it might be even weirder if we allowed `for let x := 1 lt(x, 10) x := add(x, 1) { ... }`, but this is perhaps a matter of taste.

Forcing braces for at least the first and last element introduces separators, but I'm fine with also allowing plain expressions there if you think that is useful.